File: Challenge-003-Stack/python/main.py
#!python3
class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if self.is_empty():
            print("Stack underflow")
            return None
        return self.items.pop()

    def peek(self):
        if self.is_empty():
            print("Stack is empty")
            return None
        return self.items[-1]

    def is_empty(self):
        return len(self.items) == 0

# Ejemplo de uso
if __name__ == "__main__":
    stack = Stack()
    stack.push(10)
    stack.push(20)
    stack.push(30)

    print("Cima:", stack.peek())      # 30
    print("Pop:", stack.pop())        # 30
    print("Nueva cima:", stack.peek()) # 20

File: Challenge-003-Stack/rust/src/main.rs
/* Template for Rust program */

struct Stack<T> {
    items: Vec<T>,
}

impl<T> Stack<T> {
    fn new() -> Self {
        Stack { items: Vec::new() }
    }

    fn push(&mut self, item: T) {
        self.items.push(item);
    }

    fn pop(&mut self) -> Option<T> {
        self.items.pop()
    }

    fn peek(&self) -> Option<&T> {
        self.items.last()
    }

    fn is_empty(&self) -> bool {
        self.items.is_empty()
    }
}

fn main() {
    let mut stack = Stack::new();
    stack.push(10);
    stack.push(20);
    stack.push(30);

    println!("Cima: {:?}", stack.peek()); // Some(30)
    println!("Pop: {:?}", stack.pop());   // Some(30)
    println!("Nueva cima: {:?}", stack.peek()); // Some(20)
}